Unify Chip styles

Aljullu opened the pull request:
Fixes #1786.

This PR updates the Filter Products by Attribute and Active Filters blocks so they use the `Chip` component introduced in the Cart & Checkout blocks. `Chip` needed some changes so it could be used in those scenarios.

I also took the chance to fix some theme compatibility issues in those blocks (input backgrounds & Clear All color).

We are making some style changes and switching some components, so there is the possibility of themes breakage but I think it's quite small. IMO there is no need for a dev note. Thoughts?

#### Accessibility

- [x] I've tested using only a keyboard (no mouse)
- [x] I've tested using a screen reader

### Screenshots
Screenshots with Bistro theme:

_Before:_
![imatge](https://user-images.githubusercontent.com/3616980/85393809-e65a2c80-b54d-11ea-801e-2685484459ca.png)

_After:_
![imatge](https://user-images.githubusercontent.com/3616980/85393664-aabf6280-b54d-11ea-933e-5ca448c45aba.png)

### How to test the changes in this Pull Request:
0. Ideally, set your theme background to something different from white so you can verify inputs are still legible.
1. Create a page with the All Products block and the Filter Products by Attribute (set the attributes to Query Type: OR, Display Style: Dropdown) and Active Filters (Display Style: Chips).
2. Filter Products by Attribute:
  2.1. Try adding new filters.
  2.2. Verify chips have the correct styles.
  2.3. Try removing them with the keyboard (backspace or <kbd>Del</kbd>).
  2.4. Try removing them clicking on the chip name.
3. Active Filters:
  3.1 Verify chips have the correct styles.
4. Catching regressions:
  4.1. Verify there are no regression in the Filter Products by Attribute and the Active Filters blocks with other attribute combinations: verify everything is still working and there are no visual bugs.
  4.2. Verify there are no regressions with the Chips in the Cart: try adding a coupon in the Cart or Checkout blocks and verify it still looks correct.

### Changelog

> Chip styles of the Filter Products by Attribute and Active Filters have been updated to give a more consistent experience.

nerrad commented on assets/js/base/components/chip/index.js:
```diff
@@ -25,46 +25,79 @@ const Chip = ( {
-			<button
+			{ showScreenReaderText && (
+				<span className="screen-reader-text">
+					{ screenReaderText ? screenReaderText : text }
```
It looks like this could just be `screenReaderText` now right?
```suggestion
					{ screenReaderText }
```

nerrad commented on assets/js/base/components/chip/test/index.js:
```diff
@@ -16,6 +16,14 @@ describe( 'Chip', () => {
 			expect( component.toJSON() ).toMatchSnapshot();
 		} );
 
+		test( 'should render nodes as the text', () => {
+			const component = TestRenderer.create(
+				<Chip text={ <h1>Test</h1> } />
+			);
+
+			expect( component.toJSON() ).toMatchSnapshot();
+		} );
```
We should switch to use React Testing Library here. I understand though if you want to leave this for now and revisit in a potential follow-up.

Aljullu commented on assets/js/base/components/chip/index.js:
```diff
@@ -25,46 +25,79 @@ const Chip = ( {
 	onRemove = () => {},
```
Changed in: 692fd0515740d0b4c2f26b610102f30cc6b91bb5.

Aljullu commented on assets/js/base/components/chip/index.js:
```diff
@@ -25,46 +25,79 @@ const Chip = ( {
 	onRemove = () => {},
 	disabled = false,
 	radius = 'small',
+	removeOnClick = false,
```
> So with that in mind, the prop name seems like it could use some "tightening up" so it's purpose is a bit more intuitive. Maybe something like `removeOnAnyClick`?

:+1:  Done in ce51de2117cceec4305817932a33caf36045917a.

> Also, I wonder, will there ever be any use-case for using this same Chip Pattern for something that is _not_ removable (meaning the chip is persistently displayed)? Should we differentiate the name of this particular component so it's clear it's always removable (i.e. `RemovableChip` or `DismissableChip`)? Eventually the ui/ux could be moved into `<Chip>` and this component (<DismissibleChip>) would wrap around the former enhancing it with the removal actions/labels.

Yes, that sounds good to me. Just thinking we should wait until we see the real need to have an un-removable chip before differentiating between two kinds of chips?

nerrad replied:
> Just thinking we should wait until we see the real need to have an un-removable chip before differentiating between two kinds of chips?

I'd be okay with that, but I just wonder if it would be better to be more explicit _now_ with the name of the component and call it `RemovableChip` or `DismissibleChip` which leaves room for having just `Chip` down the road. It makes it a bit clearer that this is a component with dismissing behaviour?

Aljullu replied:
I ended up splitting the component in two (`Chip` and `RemovableChip`) as you suggested in the first place. This way both components can share the same CSS classes so it's not a breaking change: d0eeb266a71d498fdcdff3f4826185335e19da93.

Aljullu commented on assets/js/base/components/chip/index.js:
```diff
@@ -25,46 +25,79 @@ const Chip = ( {
+		'aria-label':
+			ariaLabel || typeof text !== 'string'
+				? ariaLabel
+				: sprintf(
+						/* translators: %s text of the chip to remove. */
+						__( 'Remove "%s"', 'woo-gutenberg-products-block' ),
+						text
+				  ),
```
Hmm, I'm not sure to understand you here. There are three posibilities:
* `arilaLabel` prop is set: in that case, we use `ariaLabel`.
* `ariaLabel` is not set:
  * If `text` is a string, it uses `Remove "%s"`.
  * If `text` is not a string, it uses `Remove`.

That's because `text` accepts any node, so it could be a string but also an element.

nerrad replied:
> arilaLabel prop is set: in that case, we use ariaLabel.

In the code it's _always_ going to be set because the destructured prop value has a default:

```
ariaLabel = __( 'Remove', 'woo-gutenberg-products-block' ),
```

So the expression `ariaLabel || typeof text !== 'string'` will never evaluate the second condition _unless_ `ariaLabel` is explicitly set to something a non-truthy value and not `undefined` or not set. 

The problem I'm seeing might be more clear if you add some tests covering the conditions this expression uses.

Aljullu replied:
Aaaah, of course! :facepalm: I updated the logic in 273fde8cc42fe1a83467b8900cc10e7cf61f9d90. I also made it so it honors the screen reader text. I think it makes more sense to use that one than the default `text`.

> The problem I'm seeing might be more clear if you add some tests covering the conditions this expression uses.

I think it's now covered with the tests added in d0eeb266a71d498fdcdff3f4826185335e19da93.

Aljullu commented:
Thanks for the review @nerrad!

> First, the dismiss "x" for the coupon chips seems to be a bit off center vertically.

Good catch! I was struggling to get it correctly aligned. Turns out we were using the multiplication X character, which seems to be a bit misaligned to the top. I replaced it with the [cancellation X character](https://en.wikipedia.org/wiki/X_mark#Unicode) in 8edd9df09b690fc82ccddc089c14a29fca0c7bf0:

| Before | After |
| --- | --- |
| ![imatge](https://user-images.githubusercontent.com/3616980/85740495-f61a7200-b701-11ea-9850-c4699d67213d.png) | ![imatge](https://user-images.githubusercontent.com/3616980/85740447-ed29a080-b701-11ea-969f-46155878621f.png) |

At some point we could replace it with an SVG icon, but I think the character looks good enough.

> Next, the space between the chip and the bottom of the input field (with the filter block) is more than the margin at the top (and the chip doesn't look centered in it's container):

Ah, a specificity issue. :slightly_smiling_face:  Fixed in: 392e2b1734c59bd05c49c4ba6b14b10269e19469.

nerrad commented:
> At some point we could replace it with an SVG icon, but I think the character looks good enough.

Ya cancellation character definitely looks better 👍

nerrad commented:
I haven't reviewed the code yet, but did some testing. Interestingly at first with `npm run build` I got this:

Chrome:
<img width="210" alt="Image 2020-07-06 at 1 56 05 PM" src="https://user-images.githubusercontent.com/1429108/86624688-c1c07480-bf91-11ea-87ba-6b3bf7d2a289.png">

Safari:
<img width="258" alt="Image 2020-07-06 at 1 55 54 PM" src="https://user-images.githubusercontent.com/1429108/86624699-c7b65580-bf91-11ea-8cbf-657394bd8ccc.png">

So I wiped my `node_modules` folder, did a fresh `npm install` and `npm run build` and this time ended up with this:

Chrome:
<img width="418" alt="Image 2020-07-06 at 2 04 39 PM" src="https://user-images.githubusercontent.com/1429108/86624771-e7e61480-bf91-11ea-9de4-4a03b6a2446d.png">

Safari:
<img width="393" alt="Image 2020-07-06 at 2 04 50 PM" src="https://user-images.githubusercontent.com/1429108/86624788-ef0d2280-bf91-11ea-8da6-dfd420ecd303.png">

Very odd! I wonder if it's due to changes with a package being updated and me having something stale in `node_modules`? Either way, probably something to watch for when this eventually lands. Also worth noting that after running `npm install` I have a modified `package-lock.json` file locally (with a number of changes...).

So with the second, successful build, there spacing looks better on the chips _outside_ of the input but if you notice in my screenshots, the chip inside the input still has unaligned spacing on the "x".

nerrad commented:
oh hang on, it looks like I was back on the main branch for my second build 🤦 , testing again.

Aljullu commented:
> I haven't reviewed the code yet, but did some testing. With `npm run build` I got this:

That's weird, it's working fine for me. To me, it looks like an issue with the font?

In 2ded0b844b93d040507580f964fdecb54a9b2e29 I replaced the character with an `Icon`, I guess that should fix it. Can you try?

nerrad commented on assets/js/base/components/chip/test/__snapshots__/index.js.snap:
```diff
@@ -55,67 +70,104 @@ exports[`Chip without custom wrapper should render custom aria label 1`] = `
-    🗙
+    <svg
+      aria-hidden="true"
+      className="dashicon dashicons-arrow-down-alt2"
```
🤔 hmm, this is a bit surprising, shouldn't this be the `no-alt` icon in the snapshot? Actually in looking closer, it looks like we have bad classnames in the `no-alt.js` file in our icon library folder. It could be handled in a separate issue (See #2846).

Aljullu replied:
Ah, good catch! I missed that. Will open a PR today fixing it.

nerrad commented on assets/js/base/components/dropdown-selector/style.scss:
```diff
@@ -1,6 +1,6 @@
 // 17px is the minimum input field line-height and 14px is the font-size of
 // the drop down selector elements.
```
```suggestion
// 18px is the minimum input field line-height and 14px is the font-size of
// the drop down selector elements.
```
